tests/kernel/futex/src/main.c

```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <irq_offload.h>
+
+/* Macro declarations */
+#define TOTAL_THREADS_WAITING (5)
+#define STACK_SIZE (1024 + CONFIG_TEST_EXTRA_STACKSIZE)
+
+#define futex_wake_from_isr(futex) irq_offload(futex_isr_wake, futex)
+
+/******************************************************************************/
+/* declaration */
+K_THREAD_STACK_DEFINE(stack_1, STACK_SIZE);
+K_THREAD_STACK_ARRAY_DEFINE(multiple_stack, TOTAL_THREADS_WAITING, STACK_SIZE);
+
+struct k_mem_domain futex_domain;
+K_APPMEM_PARTITION_DEFINE(futex_partition);
+K_APP_BMEM(futex_partition) int timeout;
+K_APP_BMEM(futex_partition) int index[TOTAL_THREADS_WAITING];
+K_APP_BMEM(futex_partition) struct k_futex simple_futex;
+K_APP_BMEM(futex_partition) struct k_futex multiple_futex[TOTAL_THREADS_WAITING];
+
+struct k_thread futex_tid;
+struct k_thread multiple_tid[TOTAL_THREADS_WAITING];
+
+/******************************************************************************/
+/* Helper functions */
+void futex_isr_wake(void *futex)
+{
+	k_futex_wake((struct k_futex *)futex, 1);
+}
+
+void futex_wait_task(void *p1, void *p2, void *p3)
+{
+	s32_t ret_value;
+	int time_val = *(int *)p1;
+
+	zassert_true(time_val >= K_FOREVER, "invalid timeout parameter");
+
+	ret_value = k_futex_wait(&simple_futex,
+			atomic_get(&simple_futex.val), time_val);
+
+	switch (time_val) {
+	case K_FOREVER:
+		zassert_true(ret_value == 0,
+		     "k_futex_wait failed when it shouldn't have");
+		zassert_false(ret_value == 0,
+		     "futex_wait_task wakeup when it shouldnt't have");
+		break;
+	case K_NO_WAIT:
+		zassert_true(ret_value == -EBUSY,
+		     "k_futex_wait failed when it shouldn't have");
+		atomic_sub(&simple_futex.val, 1);
+		break;
+	default:
+		zassert_true(ret_value == -ETIMEDOUT,
+		     "k_futex_wait failed when it shouldn't have");
+		atomic_sub(&simple_futex.val, 1);
+		break;
+	}
+}
+
+void futex_wait_wake_task(void *p1, void *p2, void *p3)
```

Comment:
this function is nearly identical to futex_wait_task, it would help to document why they have to be separate, or better yet combine them